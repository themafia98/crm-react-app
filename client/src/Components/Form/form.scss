
form {
    display: flex;
    align-items: center;
    justify-content: center;

    .form__textInput{
        margin: 0 5px;
        font-size: 100%;
        width: 20%;
        height: 60px;
        white-space: normal;
        padding: 0 20px;
        font-size: 16px;
        box-sizing: border-box;
        line-height: 1.33;
        border: 0 none;
        border-radius: 10px;
        outline: none;
        color: #000000;
        background-color: #efefef;
    }

    .form__submit{
        cursor: pointer;
        font-size: 1.05rem;
        width: 30%;
        color: white;
        font-weight: bold;
        background-color: #db4646;
        transition: all .3s linear;
    }
    .form__submit:hover,
    .form__submit:active {
        color: black;
        box-shadow: 0 0 10px;
    }

    .form__submit:active{
        color: wheat;
    }
}
