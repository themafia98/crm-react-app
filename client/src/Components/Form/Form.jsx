import React,{useState,Fragment} from 'react';
import PropTypes from 'prop-types';
import eventEmitter from '../../EventEmitter';
import isFetch from 'isomorphic-fetch';
import './form.scss';
const Form = ({mode}) => {

    const [disabled, setDisabled] = useState(false);
    const [error, setError] = useState('');
    const [done, setDone] = useState(false);
    const refForm = React.createRef();

    
    const validate = (data) => {
        const isEmail = /\w{1,18}@\w{1,18}\.\w{1,10}/.test(data.email);
        const isName = /^\D+$/g.test(data.name);
        const isNumber = /^\+?\d{3,10}(\(\d+\))?\d{3,10}$/.test(data.number);

        if (isEmail && isName && isNumber) return true;
        else return false;
    };

    const openModalPolicy = event => {

        eventEmitter.emit('EventOpenModal', {action: 'policy', active: true});
        event.stopPropagation();
    }
    
    const sendRequest = event => {

        setDone(false);

        const form = refForm.current;

        if (!form) return;
        
        const isFull = form.email.value && form.name.value  && form.number.value;

        const isValid = validate({
            email:  form.email.value,
            name: form.name.value,
            number: form.number.value,
        });

        if (!isFull) {
            return setError('Не все поля заполнены!');
        } else if (!isValid){
            return setError('Формат формы не соблюден!');
        }

        setDisabled(true);
        let adress = 'http://localhost:3001/mail/sendMailConsultation';
        if (process.env.NODE_ENV === 'production')
        adress = process.env.REACT_APP_CONSUL;
        isFetch(adress, {
            method: 'POST',
            body: new FormData(refForm.current),
        })
        .then(res => { 
            if (res.status === 200) {
                setDisabled(false);
                setError('');
                setDone(true);
            }
            else if (res.status === 400) 
                throw new Error('Ошибка отправки запроса!');
        })
        .catch(error => {
            setDisabled(false);
            console.error(error.message);
            setError(error.message);
        });
    }

    const sendRequestQuestion = event => {

        setDone(false);

        const form = refForm.current;
        const isFull = form.email.value && form.name.value  && form.number.value && form.text.value;
        
        const isValid = validate({
            email:  form.email.value,
            name: form.name.value,
            number: form.number.value,
        });

        if (!isFull) {
            return setError('Не все поля заполнены!');
        } else if (!isValid){
            return setError('Формат формы не соблюден!');
        }
        setDisabled(true);
        let adress = 'http://localhost:3001/mail/sendMailQuestion';
        if (process.env.NODE_ENV === 'production')
        adress = process.env.REACT_APP_QUESTION;
        isFetch(adress, {
            method: 'POST',
            body: new FormData(refForm.current),
        })
        .then(res => { 
            if (res.status === 200) {
                setDisabled(false);
                setError('');
                setDone(true);
                console.log('Request send!');
            }
            else if (res.status === 400) 
                throw new Error('Ошибка отправки запроса!');
        })
        .catch(error => {
            setDisabled(false);
            console.error(error.message);
            setError(error.message);
        });
    }

    if (mode === 'index'){
        return (
            <section key = 'fromFeedBack' className = 'feedBack container'>
                <form className = {mode} ref = {refForm}>
                    <input 
                        name = 'email'  
                        className = 'form__textInput' 
                        type = 'email' 
                        placeholder = 'E-mail' 
                    />
                    <input 
                        name = 'name'  
                        className = 'form__textInput' 
                        type = 'text' 
                        placeholder = 'Имя' 
                    />
                    <input 
                        name = 'number' 
                        className = 'form__textInput' 
                        type = 'tel' 
                        placeholder = 'Телефон' 
                    />
                    <input
                        onClick = {sendRequest}
                        disabled = {disabled}
                        className = 'form__textInput form__submit' 
                        type = 'button' 
                        value = 'Бесплатная консультация' 
                    />
                    {done && <span className = 'done'>Письмо отправлено, мы скоро ответим!</span>}
                    {error && <span className = 'error'>{error}</span>}
                </form>
            </section>
        )
    } else if (mode === 'contact'){
        return (
            <form key = 'formContact' className = {mode} ref = {refForm}>
                <input 
                    name = 'name' 
                    type = 'text' 
                    placeholder = 'Имя' 
                />
                <input 
                    name = 'email' 
                    type = 'text' 
                    placeholder = 'E-mail' 
                />
                <input 
                    name = 'number' 
                    type = 'text' 
                    placeholder = 'Телефон' 
                />
                <textarea 
                    name = 'text'  
                    type = 'text'
                    placeholder = 'Ваш вопрос' 
                />
                {done && <span className = 'done'>Письмо отправлено, мы скоро ответим!</span>}
                {error && <span className = 'error'>{error}</span>}
                <input
                    onClick = {sendRequestQuestion}
                    disabled = {disabled}
                    type = 'button' 
                    value = 'Отправить' 
                />
                <p className = 'policy contact-text--center'>
                    Нажимая на кнопку, вы даете согласие 
                    на обработку персональных данных 
                    и соглашаетесь c <span
                        onClick = {openModalPolicy} 
                        className = 'policy'>
                         политикой конфиденциальности
                    </span>
                </p>
            </form>
        )
    } else return <Fragment></Fragment>
};

Form.propTypes = {
    mode: PropTypes.oneOfType([
        PropTypes.string.isRequired,
        PropTypes.bool.isRequired
    ])
};

export default Form;