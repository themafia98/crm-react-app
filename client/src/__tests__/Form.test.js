import React from 'react';
import renderer from 'react-test-renderer';

import Form from '../Components/Form/Form';

const FormTest = renderer.create(
    <Form mode = 'index' />
);

test('Form test', () => {

    let treeSnapshotForm = FormTest.toJSON();
    expect(treeSnapshotForm).toMatchSnapshot();

    const buttonSend = FormTest.root.find(button => button.props.className === 'form__textInput form__submit');
    buttonSend.props.onClick({stopPropagation: () => null});

    treeSnapshotForm = FormTest.toJSON();
    expect(treeSnapshotForm).toMatchSnapshot();
});