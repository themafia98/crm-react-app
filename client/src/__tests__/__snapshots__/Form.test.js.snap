// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form test 1`] = `
<section
  className="feedBack container"
>
  <form
    className="index"
  >
    <input
      className="form__textInput"
      name="email"
      placeholder="E-mail"
      type="email"
    />
    <input
      className="form__textInput"
      name="name"
      placeholder="Имя"
      type="text"
    />
    <input
      className="form__textInput"
      name="number"
      placeholder="Телефон"
      type="tel"
    />
    <input
      className="form__textInput form__submit"
      disabled={false}
      onClick={[Function]}
      type="button"
      value="Бесплатная консультация"
    />
    
  </form>
</section>
`;

exports[`Form test 2`] = `
<section
  className="feedBack container"
>
  <form
    className="index"
  >
    <input
      className="form__textInput"
      name="email"
      placeholder="E-mail"
      type="email"
    />
    <input
      className="form__textInput"
      name="name"
      placeholder="Имя"
      type="text"
    />
    <input
      className="form__textInput"
      name="number"
      placeholder="Телефон"
      type="tel"
    />
    <input
      className="form__textInput form__submit"
      disabled={false}
      onClick={[Function]}
      type="button"
      value="Бесплатная консультация"
    />
    
  </form>
</section>
`;
